### ISSUE TITLE
Autogenerate TestMethods with NUnit 3

### ISSUE THREAD

**Issue by Stefan75** [nunit/nunit#616]
Hello,
I try to automatically create Test-Methods and add them to the manually created one.
(maybe you can call it aspect-oriented-testing)

Reason:
For a set of classes under tests, some aspects can be tested automatically.
Example:
ViewModels: They should expose public properties. Each setter should raise the correct change event.

For each public property of the class-under-test I want to add several Test-Methods:
1. `<PropName>_<CalledGetter>_ReturnsValidValue`
2. `<PropName>_<CalledSetter>_ValueSet`
3. `<PropName>_<CalledSetterWithDifferentValue>_ChangeEventRaised`
3. `<PropName>_<CalledSetterWithCurrentValue>_NoChangeEventRaised`

Usage:
```
[TestFixture]
[AutoViewModelTests]
[TestOf(typeof(ViewModelX))]
public class ViewModelXTests
{
    [Test] public void ManualTest() {}
}
```

In visual studio all tests (manually and automatically created) should looks equal.
(Every test-method should be listed as own test)

I started a prototype based on NUnit 3.
My prototype is based on an attribute with the interface `IFixtureBuilder`
```
public class AutoViewModelTestsAttribute : FixtureBuilderAttribute, IFixtureBuilder
```
In the `BuildFrom` method I created a `TestSuite` with several created `TestMethod` instances.

The visual studio test explorer (with the installed nunit adapter) shows all manually created tests but just one of the dynamically created (in the overview - in the details, I can see also the other dynamically created tests).

But already at this stage I faced the following problems:
1. I am using `Action` instances for the individual test logic. It seems that they share the same `MethodInfo` which is used in the `TestMethod` constructor. This leads to the situation that the test cases are not visible for its own - it seems that they are treated as a group and just the first one is displayed normally.
2. The `MethodInfo` objects of my actions are called with the test-class instance - which is not correct and leads to an exception `System.Reflection.TargetException : Object does not match target type.` 

Therefore it seems that I have several challenges:
1. Because my created test methods do not belong to the test-class self they need to be called as an `Action` with the test-class instance as argument.
2. There is something like a auto-grouping based on the MethodInfo object. Is it possible to disable this behavior?

Based on the current situation I want humbly ask:
1. Is it possible at all?
2. and is the current way the right one?

Solution idea:
It seems that the `TestMethod` class is based on `MethodInfo` which leads to some problems.

Basically I just want to provide a custom-logic which should be treated as a normal test-method and just be executed.

Maybe I need an alternative to `TestMethod` which is based on an `Action` ?
`var tm = new TestMethodBasedOnAction( testInstance => { testInstance.Foo()}, testSuite);`

The content of the provided `Action` is generated. During test, the action is called with the already instantiated test-class as argument. 
In the action I have some arbitrary logic which calls the right methods on the test-class
(The Test-Class implements an interface to provide the Class-Under-Test instance).

I would appreciate all kinds of feedback...

With best regards,
Stefan

**Comment by CharliePoole** [comment]
Hello Stefan,
I will initially treat this issue as a 'question' since you haven't asked us to do anything. If we find that NUnit needs to add an interface or make some other changes, then I'll re-label the issue.

First I want to restate what I understand, please correct me if necessary...

You want to create an attribute that 
* Creates a fixture or suite.
* Adds standard tests to the suite for a subject class with a certain interface.
* Also adds normal NUnit tests.

I'm especially interested in the last item. Is this a deliberate intention on your part or an accident of the implementation. It seems difficult for your class to keep up with all the types of tests that NUnit may add inside a normal TestFixture in the future unless it also shares some implementation with NUnit itself. Can you clarify what you want here?

I'm going to a meeting now, so I'll send this and come back later with some questions I have about your implementation.

Charlie

**Comment by Stefan75** [comment]
Hello Charlie,
thanks for the fast reply and sorry for my poor description. I try to explain it as good as possible.
Yes - currently it is just a question and after you understand my goal it is up to you to re-label the issue.

Related to your question:
My main intention is to "just" add tests (aka testmethods).based on several domain related aspects.
Currently the only way I found was to create a class attribute with the `IFixtureBuilder` interface (because I looked into the `TestFixture` implementation). But it is just my unskillful solution not my intend.

If my solution does not work nicely with the normal tests then it is an accident and not my intend.
I would actually say that I have chosen the wrong way!

Butt currently it seems that it is not as bad as it sounds because my test class looks currently:
```
    [AutoViewModelTests]
    [TestFixture]
    [TestOf(typeof(Foo))]
    public class FooTests {}
```
Based on the attribute usage I expect that there are actually two TestSuits created:
1. The normal one from  `TestFixture`
2. The "custom" one from `AutoViewModelTests`

It seems that the Visual Studio test runner uses both fixtures together. It seems that he merges them "silently" together. But also here: Two test fixtures was not my intend, it is just my current solution.

My overall goal is more a composing style:
I have a test class which is decorated by several "test" related attributes (e.g. TestFixture, AutoViewModelTests).
Each of them has the responsibility to provide 0..n tests.
The normal one 'TestFixture' creates for each public method with the `Test`attribute.one test.
Other attributes are using different mechanisms to create tests. At the end they should be "merged" and executed.

If I look at the results I don't care how the tests are defined and what was there origin.
I believe that for several classes it make not sense to create all tests manually and it also make no sense to create all tests automatically.

Therefore I am searching a way to have both:
I want a test class which is responsible to test a class-under-test where the actual tests are composed by different test-providers (manual and automated). 

With best regards,
stefan

**Comment by CharliePoole** [comment]
Yes, for each TestFixtureBuilder attribute you place on a test, you create a separate fixture. The TestFixture attribute creates one with all the test methods in it and your new attribute would create one with all the automatic tests. Because of how VS interfaces with NUnit, you don't see the two different suites. VS organizes tests by class, not by fixture. It doesn't even understand the NUnit concept of a fixture, only tests (methods).

For that reason, you should always develop enhancements to NUnit running in the console or gui. In this case, a gui would be best... but of course we don't have an NUnit 3.0 gui yet so that's impossible. You will have to examine the output XML to see what fixtures and tests are actually created. I recommend running at least part of the time in the console so you can see that.

Having separate fixtures is not such a bad thing, in my opinion. In my coaching, I treat the pattern of one test class for each tested class as a beginner's approach. A better organization is to have one fixture to test each different aspect or feature of the system. So a fixture dedicated only to the validation of properties makes a lot of sense.

The ITestFixtureBuilder interface is responsible for building the entire fixture, including any tests contained in it. If you are happy to have a separate fixture for automatic tests, it's a good approach. If you prefer to allow different kinds of tests being generated in the same fixture, then we would need to create something like an IAddTestsToFixture interface. That's indeed possible, but we won't do it without a strong need.

My advice is to _not_ try to combine the different kinds of tests in a single fixture. NUnit already creates a suite that contains both fixtures, so that gives you your composability. That is, your example should show up in the NUnit results as

```
ViewModelXTests
    ViewModelXTests // Your manual tests
    YourName            // Your automatic tests
```

This gets us down to the real question: How to make your automatic tests work.

First, some questions:
* How does your attribute know what view class it is supposed to be testing?
* When do you instantiate that class in the generated tests?
* You say you are using Actions. Is there a reason you prefer that to methods of the attribute?

Charlie

**Comment by Stefan75** [comment]
thanks for all the information and also the advice to separate the different aspects into different test-classes.

Now to your questions:
* I have several ideas how the find the class to be tested:
  * Maybe by an attached attribute on the test-class (e.g. `TestOf`)
  * Maybe via an interface of the test-class
  * Maybe as argument of the test attribute self (e.g. `AutoViewModelTest(typeof(Foo))`)
* In my dream I would like that nunit is still responsible to create the test class instances (like in all other tests). 
The test-class could implement an interface with a method to provide the instance of the class-under-test (e.g. `CreateCuT()`). This would also support complex creation logic with fakes.
(very similar to Example 3 in the Action-Attribute section (action attribute is applied to interface ... to provide data to tests))

* I am not sure if I understand your last question correctly. Why I want to use Actions and not method of the attribute?
   In my example use-case: Some custom logic in my attribute (placed at the class) would analyze the class-under-test type related to all public properties. Each property should be tested by some test-methods.
   An implementation could be something like:
   `relevantPublicProperty.Select(prop => CreateTest(prop)) ...`
   `CreateTest` would create a `TestMethod` with some logic (e.g. calling the given property)
    It is handy to create an Action with the code-fragment together with the captured prop argument.
     Later during execution, the action needs the actual instance of the test-class.

  I guess that currently the MethodInfo is called directly on the created test-class instance.
  In my use-case I would like an Action (with custom logic and some captured data) which is called
  with the required data (e.g. TestContext to get the created test-class instance).

  I am not pretty sure but it seems for me that such an Action based solution could also be the base for
a MethodInfo solution where the MethodInfo instance is captured (very similar to a property info).


stefan

**Comment by CharliePoole** [comment]
Finding the class to test:
 * Using TestOf seems risky. I can write `[TestOf("Creation function")]` The Type argument is just shorthand for a string and the attribute is only intended for use in reporting.

 * An interface on the test class, which is known to your attribute, is possible. It's a bit messy, I think. You have to change the internals of your test class in order to test a new Type.

 * Giving your fixture-building attribute a type argument makes it work more like other parts of NUnit. Personally, that's how I would go first.

Instantiating the test class:
 * NUnit prepares to execute a test by creating a chain of Commands to run. See CommandBuilder.cs.
 * Instantiating the object under test (fixture) is done in the OneTimeSetupCommand.
 * You need a good understanding of commands and how they are created and run to make a non-standard test fixture. :-(

Instantiating the class under test:
 * NUnit doesn't do this. Your test class should do it, based on knowledge of how to initiate such a class.
 * If pass a Type argument to the test class, you can instantiate it yourself but you would need knowledge of any initialization.

Use of Actions:
 * OK, I was thinking you had a static list of tests to run. You can't do that if you are analyzing classes on the fly and deciding what to test. This is not something NUnit does, primarily because NUnit was created by guys who think you should write your own tests. :-)
 * It may be possible to create a test fixture with _one_ test method that takes an action as a parameter. NUnit would be OK with that. However, the difficulty is that you would be providing the Action at the time the tests are loaded. It would need to remain avaiable until the run-time of the test. I'm not sure how you could do that.

**Comment by Stefan75** [comment]
It seems providing the Type via the Attribute is the best solution and fits nicely to the NUnit philosophy.
=> Thanks for the feedback

Related to instantiation and execution:
After looking carefully to the sources I understand your comments:
- WorkItem CreateWorkItem(iTest, ITestFilter) with `SimpleWorkItem((TestMethod)test)`
- SimpleWorkItem Constructor(TestMethod) with `CommandBuilder.MakeTestCommand`
- CommandBuilder MakeTestCommand(TestMethod) with `new TestMethodCommand(test)`
- TestMethodCommand `Execute` and `RunNonAsyncTestMethod`

The current design of these parts is based on the fact that there are just two `ITest` implementations:
1. TestSuite which leads to a CompositeWorkItem
2. TestMethod which leads to a SimpleWorkItem

CommandBuilder MakeTestCommand knows exactly that a `TestMethodCommand` is needed for the `ITest` implementation `TestMethod`.

TestMethodCommand running is based on Reflect.InvokeMethod and the knowledge that the `TestMethod` object has a property with the name `Method`. 

=> With the current design it is not easy to create an additional implementation of ITest similar to TestMethod because the framework knows and uses the concrete types. It knows during the transformation of ITests to WorkItems/Commands which types are expected and how they are mapped/related.

Is the summary correct?


Currently I come to the conclusion that realizing my goals with maximum reuse of nunit 
would require that these parts are changed in a way that providing other ITest implementations is intended. Which would also mean that the actual running should be moved to the Test self (ITestMethod interface - maybe with a changed name `IRunnableSimpleTest`)

Then it would be possible that `TestMethodCommand` would just depend on that interface `IRunnableSimpleTest`. This interface would provide methods like `RunNonAsync(context, arguments)`.  Then it would be possible to create an implementation based on Actions but still reuse nearly everything of NUnit (e.g. WorkItems, Command, Setup, Teardown, Actions, ...).

=> I am not sure that my goals are worth these or other changes

What do you think?

I like very much the concept of Actions based on attributes and interfaces. It allows a very flexible and composable way of adding extra functionality. The concept is based on the consistent usage of interfaces.

I don't have a full picture but maybe it is also possible to use this philosophy in this part of nunit (CompositeWorkItem vs SimpleWorkItem seperation just based on ITest derviced Interfaces, TestMethodCommand constructor based on an interface parameter, move of the actual running, ...)

Now to the other topics:
The class-under-test should be created by the test-class and provided via interface.
(This was also the way I want to go - thanks to your good examples related to Actions)

It is not my intend to mess up with the notion of TestFixture or Suite. If my intend does not fit nicely I will stop trying it. 
I "just" want
- to replace the logic how TestMethods are found (not just based on test-class Methods)
- that iTest  (TestMethod) is not strictly based on MethodInfo (e.g. with Actions)
- that ITest (TestMethod) could also dynamically created (e.g. with Actions)


stefan

**Comment by CharliePoole** [comment]
NUnit, like almost all the so-called "xunit" frameworks, is based on the Composite Pattern. That is, it treats tests and groups of tests identically through an interface (ITest). This is not a limitation - it's natural that every test is either a single test (test case) or a group (suite).

There is no limitation on the number of implementations of each of these two. NUnit has six or seven implementations of a test suite, for example. Some of those represent the test (suite) by a class, some by a method. Currently, NUnit only contains one implementation for a single test: TestMethod. That is what makes it appear to you that a method is required, but it actually isn't.

Almost all parts of NUnit will handle tests no matter how they are implemented. You just have to provide the implementation. But of course, since we never have had any single test cases except for TestMethods, some errors have crept in. One of these is that the CommandBuilder class "knows" that it must use a TestMethodCommand to execute a single test case.

This is fixable. Originally, each test created it's own command hierarchy. This seemed to me to be too much knowledge in the test about how it would be executed. [A principle of the new NUnit 3.0 design is that tests _no longer execute themselves_ but are able to be executed in a variety of ways.

Test attributes may, however, contribute to the execution of the tests on which they are placed by supplying commands. Right now, there is no interface that allows an attribute to replace the fundamental test execution command around which all others are wrapped. We can add such an interface if it proves to be needed for this kind of extension.

So, your statement...

"With the current design it is not easy to create an additional implementation of ITest similar to TestMethod because the framework knows and uses the concrete types. It knows during the transformation of ITests to WorkItems/Commands which types are expected and how they are mapped/related."

is correct, but the knowledge is not as widespread as you may think. Originally each test object type (TestMethod, TestFixture, etc.) provided it's own command structure through a MakeTestCommand method. I don't think I want to go back to that approach, but we can give attributes more opportunities to contribute to the command structure.

Moving to your suggestions:

1. Have the test execute itself. This was a fundamental problem with NUnit 2.x as it applies to parallel execution. If tests execute themselves, then it's difficult to have a bunch of separate worker threads that execute different parts of the test. That's why we went to the approach of first breaking the requirements of test execution into a command structure. Attributes already contribute to that command structure through IWrapTestMethod and IWrapSetUpTearDown.

2. The changes you want to make are definitely "worthy" of adding support in NUnit. The intent is to make it possible to implement tests in a variety of ways. You are the first person to attempt to create a test case that is not implemented by a method. I don't think you will be the last. That said, it will take a lot of work on your part to do it as well as a small amount of our effort. The difficulty will be to come up with a clear definition of an interface we can implement. There may also be some assumptions made in the commands that wrap TestMethodCommand about what they are wrapping. But if there are, those are simply bugs to fix.

Here is what I think is needed to implement your agenda:

1. Create an ITest-derived class that represents what you want to have done. Ideally, you should derive it from Test but that's not absolutely necessary. In place of a method, it will need to have an Action. Creating the action should not use any resources, bearing in mind that NUnit will create your class when it _loads_ the test, not when it runs it. I believe you will be able to test this class independently of the remaining steps.

2. Create a TestActionCommand, similar to TestMethodCommand. It should simply know how to execute your special test type, by calling it's action. [We may need to rename this, but that's a good name for now.]

3. Some way to get your test to be added to the fixture and your commands to be executed. This is the key problem because NUnit looks for an attribute on the object that has the tests - i.e. the method. But there is nothing in your code on which to place an attribute, except for the suite. I'll need to think about this a little to come up with an approach.

Charlie

**Comment by Stefan75** [comment]
Thanks again for the detailed answer. I want to emphasize that I don't have a clue about NUnit or other "xUnit-style" test frameworks. So please be lenient towards me.

Related to the agenda:
I will download the sources of NUnit and will create a ITest derived class as described. The same also for the TestActionCommand (I agree with the remarks related to the name :) ). 

In this first prototype I will create the "connection" somewhere in the nunit code because it is possible
that more surprises are waiting for me. If the prototype behaves in principle as desired we can continue on the other open points (if it is really worth the effort).

Thanks again for the open and fast discussion and the detailed answers. 

Stefan

**Comment by appel1** [comment]
I would find enabling non TestMethod ITest implementations useful as well. To simplify tests of certain aspects we've created a simple attribute based solution. 

Currently we've had to make it mandatory to apply these attributes on a method with a specific signature and implementation so that we have a method for the test to run. But it would be a little bit less error prone if we could set it on a property or the fixture class instead. 

```csharp
        [EqualityContractTest(typeof(PointInTime), nameof(EquivalenceClasses), ImplementsOperatorOverloads = true)]
        public void EqualityContract(TestSpecification specification)
        {
            specification.Execute();
        }
```
